### ISSUE TITLE
Request: Support next.js 9.3

### ISSUE THREAD

**Issue by LassiterJ** [lfades/next-with-apollo#114]
Would like an update for the new version of next.js. Specifically including the new data-fetching methods and examples of use.

**Comment by lfades** [comment]
There's no need for this package, or of any Apollo packages at all if you want to enjoy the goodies of the latest release of Next.js.

For `getStaticProps` do a simple fetch for the data using a GraphQL query (you can use `node-fetch` or `@zeit/fetch`) and return it as props, there's no need to add react-apollo for this case.

**Comment by LassiterJ** [comment]
> There's no need for this package, or of any Apollo packages at all if you want to enjoy the goodies of the latest release of Next.js.
> 
> For `getStaticProps` do a simple fetch for the data using a GraphQL query (you can use `node-fetch` or `@zeit/fetch`) and return it as props, there's no need to add Apollo for this case.

Ok thank you. I wasn't sure if the update would change anything with this package. Especially since its deprecates getInitialProps. Could be nice to mention in docs... But anyways thanks for responding!

**Comment by lfades** [comment]
@LassiterJ `getInitialProps` is not really deprecated and won't go away anytime soon (I'm part of the Next.js team), the package will stay as it is for now, once we see enough traction for SSG I'll update it.

**Comment by juanpablocs** [comment]
add support with `getServerSideProps` please
```js
export const getServerSideProps = async ({ query, res }) => {
  const { slug } = query;
  if (!slug && res || slug.length < 2 && res) {
    res.statusCode = 404
  }
// it's necesary for display 404 custom error
  const { data, error } = useQuery(MOVIE_SLUG, { variables: { slug } }); // not work....
  if (data && data.movie === null) {
    res.statusCode = 404;
  }
 ...
};

```

**Comment by hoangvvo** [comment]
@juanpablocs That is not how you suppose to use `useQuery` hook. A React hook can only be used in a React component. `getServerSideProps` is like an HTTP API endpoint and is definitely not React.js.

**Comment by hoangvvo** [comment]
I imagine it would be difficult to make `next-with-apollo` to work with `getServerSideProps` because `getServerSideProps` is completely separated from the React component, so there is no way to connect `getServerSideProps` to the the component (to restore cache / apolloState, `apollo` props).

**Comment by hoangvvo** [comment]
This would be how we may use Apollo Client in `getServerSideProps` and `getStaticProps`. This is modified from the example.

Export `clientFn` (first arg in `withApollo`):

```
// lib/withApollo.js
import withApollo from 'next-with-apollo';
import ApolloClient, { InMemoryCache } from 'apollo-boost';
import { ApolloProvider } from '@apollo/react-hooks';

export const clientFn = ({ initialState } = {}) =>
  new ApolloClient({
    uri: 'https://mysite.com/graphql',
    cache: new InMemoryCache().restore(initialState || {}),
  });

export default withApollo(clientFn, {
  render: ({ Page, props }) => {
    return (
      <ApolloProvider client={props.apollo}>
        <Page {...props} />
      </ApolloProvider>
    );
  },
});
```

```
// pages/index.js
import gql from 'graphql-tag';
import { useQuery } from '@apollo/react-hooks';
import withApollo, { clientFn } from '../lib/withApollo';

const QUERY = gql`
  {
    title
  }
`;

const Index = () => {
  const { loading, data } = useQuery(QUERY);

  if (loading || !data) {
    return <h1>loading...</h1>;
  }
  return <h1>{data.title}</h1>;
};

export const getServerSideProps = async () => {
  const apolloClient = clientFn();
  // apolloClient.* https://www.apollographql.com/docs/react/api/apollo-client/#apolloclient-functions
  // You do not use hooks like useQuery, useMutation in here
  return {
    props: {
      apollo: apolloClient,
      apolloState: {
        data: apolloClient.cache.extract(),
      },
    },
  };
};

export default withApollo(Index);
```

**Comment by pnfcre** [comment]
@hoangvvo Hi there, I tried your approach, but I'm getting:

```
Error serializing `.apollo` returned from `getServerSideProps` in "/". Reason: `object` ("[object Object]") cannot be serialized as JSON. Please only return JSON serializable data types.
```

Just curious to know if you can find the issue 😅 Thanks!

**Comment by lfades** [comment]
@hoangvvo You're not wrong, hence why we haven't updated the Apollo examples in the Next.js repo, but that surely will change with time, in the mean time is okay to use `getInitialProps`, it's not going anywhere

**Comment by vmptk** [comment]
@lfades , have you tried using the beta version of the @apollo/client with nextjs 9.3+?

**Comment by lfades** [comment]
I haven't, is there something new that requires an update?

I still don't like the size of their package, keeps getting bigger: https://bundlephobia.com/result?p=@apollo/client@3.0.0-beta.44 :fearful:

**Comment by wesbos** [comment]
I just tried it with Next 9.3 and it seems to work. I do see loading state for some of my content when I view source though - unsure if that is related?

**Comment by webcoderkz** [comment]
@wesbos going to try Next 9.3.5 / TS with this package.

**Comment by lfades** [comment]
@wesbos SSR is now disabled by default unless you manually enable it:
```jsx

**Comment by samuelcastro** [comment]
Why are you disabling it by default?

And does this lib mirror the official `with-apollo` next.js example? I'm using that example but I think I should just get rid of it and use this one.

Thanks!

**Comment by lfades** [comment]
@samuelcastro Feel free to use that example of this lib, you can't go wrong with either (I also helped trying to keep that example updated).

SSR is disabled by default because static pages are faster, better overall, and more reliable, and can also be dynamic when you take advantage of [SSG](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation), it just happens that it's harder to do with the current state of the Apollo library. But that surely will be improved in the future.

**Comment by webcoderkz** [comment]
@lfades should we use this: https://github.com/zeit/next.js/tree/master/examples/with-apollo, or your lib? I've noticed they've added @apollo/react-ssr library

**Comment by lfades** [comment]
@webcoderkz Not really, `getServerSideProps` is better in the way that it's never executed in the client, so it's only included for the server bundle.

The reason I haven't migrated the library yet is because `getInitialProps` is a static method inside the page, and you can more easily add an Apollo HOC on top of your pages, If I change it to work with `getStaticProps` it would take additional steps to get the setup working.

**Comment by CharlieDieter** [comment]
This question kind of extends beyond just the use of this library or Apollo, but I'm curious to know: is there a 'best practice' way to do what @juanpablocs mentioned above for dynamically returning a status code from the server using Apollo? I want to return an appropriate status code from the server based on what exists in a CMS and could just use fetch, but I would like to keep Apollo in some places in order to `fetchMore` on the frontend.   Any advice would be greatly appreciated, and apologies if this isn't the right place for this.

**Comment by batusai513** [comment]
After a few hours trying to make this work with `@apollo/client@3.0.0-beta.44`, I can see that the actual implementation and examples are kind of working, with `next-with-apollo`, I'm always seeing the loading state on the client and never getting the data hydration data, I tested both app and per page HOC, and with the example from [the next.js page with apollo](https://github.com/zeit/next.js/tree/canary/examples/with-apollo), it works but I have to explicitly add the `ssrMode: true` always, for the server and the client, in order to not see the loading state, otherwise I can see the loading, but the hydration data is on the rendered Html.

**Comment by lfades** [comment]
@CharlieDieter Why do you need to return a status code from the server to the client?

Overall I would like to just completely change the implementation of Apollo in the Next.js examples as it keeps breaking constantly for some reason :cry:

**Comment by CharlieDieter** [comment]
In my case, I have a CMS backend where content creators can build pages with a pathname. So if you go to `/example`, the page at `/pages/[uid].js` will make a graphql request for a document with `uid: example`.  That's fine, but then if you go to `/asdfasdf` and no page is found by `uid: asdfsdf` in the CMS, the CMS will return an error and that should return a 404 not found error to the client. Another possible example of wanting to change the status code could be to do a 301 or 302 redirect.

**Comment by SoorajChandran** [comment]
> I just tried it with Next 9.3 and it seems to work. I do see loading state for some of my content when I view source though - unsure if that is related?

Looks like it is related. I'm having the same issue. When viewing source I see the loading state.

**Comment by SoorajChandran** [comment]
I'm trying to make `getServerSideProps` work with apollo with the following code(based on @hoangvvo 's suggestion)

```

interface Props {
  userFromServer: User
}

const UserProfilePage: React.FunctionComponent<Props> = ({ userFromServer }) => {
  const router = useRouter()
  const { data, loading, error } = useQuery(USER_QUERY, {
    variables: { userName: router.query.userId },
  })

  if (error) {
    return (
      <Layout title="Something">
        <div className="mt-4">
          <PageLoading />
        </div>
      </Layout>
    )
  }

  if (loading) {
    const user: User = userFromServer
    return (
      <Layout
        title={`Something`}
        description={`${user ? user.about : ''}'s profile`}
      >
        {user && <UserDetail user={user} />}
      </Layout>
    )
  }

  if (data) {
    const user: User = data.user
    return (
      <Layout
        title={`${user ? user.name : ''}'s profile`}
        description={`${user ? user.about : ''}`}
      >
        {user && <UserDetail user={user} />}
      </Layout>
    )
  }

  return null
}

export const getServerSideProps: GetServerSideProps = async () => {
  const apolloClient = clientFn()

  const response = await apolloClient.query({
    query: USER_QUERY,
    variables: { userName: 'sooraj' },
  })

  const userFromServer = response.data.user
  return { props: userFromServer }
}

export default UserProfilePage
```

In the `getServerSideProps` function the data is being fetched correctly. I logged it and I can see the data. However, when I pass it as props to the actual component, the data is always undefined. 

When I view source I'm seeing the whole data rendered as JSON under and object `{"props":{"pageProps":{`

I'm new to Next.js - what am I missing here?

**Comment by km-tr** [comment]
@SoorajChandran 
Isn't this a different implementation from @hoangvvo's proposed one that passes client and apolloState, since it passes the value obtained by SSR directly as props?

**Comment by ITSWYoo** [comment]
@SoorajChandran  hi
i use graphql-code-generator
do you known how to generate this code to typescript code ?
```js
 const response = await apolloClient.query({
    query: USER_QUERY,
    variables: { userName: context.params?.userId },
  })
```
i want to query, variables type checking

**Comment by aamorozov** [comment]
I'm currently evaluating the options to move from `getInitialProps` to either full `getStaticProps`(SSG) or a hybrid and based on what I've read in this thread it doesn't seem to be needed? I'm currently setting up the apollo client and then using it in pages to query the data at build time for full SSG and it seems to work, but i'm not able to use cached queries - posted a discussion here if you're interested @lfades :
https://github.com/vercel/next.js/discussions/14471
